File: 2089-maximum-matrix-sum/maximum-matrix-sum.py
class Solution:
    def maxMatrixSum(self, matrix: List[List[int]]) -> int:
        # pass negative sign to the smallest value
        n = len(matrix)
        num_neg, smallest_value = 0, 10 ** 5
        matrix_sum = 0

        for r in range(n):
            for c in range(n):
                num = matrix[r][c]
                smallest_value = min(smallest_value, abs(num))
                if num < 0:
                    num_neg += 1
                matrix_sum += abs(num)

        # print(matrix_sum, smallest_value)
        return matrix_sum if num_neg % 2 == 0 else matrix_sum - 2 * smallest_value

File: 2089-maximum-matrix-sum/maximum-matrix-sum.java
class Solution {
    public long maxMatrixSum(int[][] matrix) {
        // migrate the negative sign
        int n = matrix.length;
        int num_neg = 0;
        long total = 0;
        int min_abs_value = Integer.MAX_VALUE;

        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                int num = matrix[r][c];
                if (num < 0) num_neg++;
                min_abs_value = Math.min(min_abs_value, Math.abs(num));
                total += Math.abs(num);
            }
        }

        if (num_neg % 2 == 0 || min_abs_value == 0) return total;
        return total - 2 * min_abs_value;
    }
}